llvm/lib/Transforms/Yk/ModuleClone.cpp

```diff
@@ -0,0 +1,78 @@
+#include "llvm/Transforms/Yk/ModuleClone.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Module.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Linker/Linker.h"
+#include "llvm/Pass.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+
+#define DEBUG_TYPE "yk-module-clone-pass"
+
+using namespace llvm;
+
+namespace llvm {
+void initializeYkModuleClonePass(PassRegistry &);
+} // namespace llvm
+
+namespace {
+struct YkModuleClone : public ModulePass {
+  static char ID;
+  const std::string ClonePrefix = "__yk_clone_";
+
+  YkModuleClone() : ModulePass(ID) {
+    initializeYkModuleClonePass(*PassRegistry::getPassRegistry());
+  }
+
+  void updateClonedFunctions(Module &M) {
+    for (llvm::Function &F : M) {
+      if (F.hasExternalLinkage() && F.isDeclaration()) {
+        continue;
+      }
+      F.setName(ClonePrefix + F.getName().str());
+    }
+  }
+
+  void updateClonedGlobals(Module &M) {
+    for (llvm::GlobalVariable &GV : M.globals()) {
+      std::string GlobalName = GV.getName().str();
+      if (GlobalName.find('.') == 0) {
+        continue; // This is likely not user-defined. Example: `.L.str`.
+      }
+      GV.setInitializer(nullptr); // Remove the initializer
+      GV.setLinkage(llvm::GlobalValue::ExternalLinkage); // Set external linkage
+    }
+  }
+
+  bool runOnModule(Module &M) override {
+    if (M.getName() == "src/perf/collect.c") {
+      return false;
+    }
+    std::unique_ptr<Module> Cloned = CloneModule(M);
+    if (!Cloned) {
+      errs() << "Error: CloneModule returned null for module: " << M.getName()
+             << "\n";
+      return false;
+    }
+    updateClonedFunctions(*Cloned);
+    updateClonedGlobals(*Cloned);
+
+    llvm::Linker TheLinker(M);
+    if (TheLinker.linkInModule(std::move(Cloned))) {
```

Comment:
Ah hah! So you found a bit to do the merging for you. What exactly does this call do?

Reply:
This call attempts to link/merge the both module `std::move(Cloned)` moves ownership of the cloned module to the linker, meaning that after this call the cloned module is no longer valid in the original context.
If the modules are successfully linked, the function returns false (curious choise). If there's an error (such as conflicting function or global names that cannot be resolved), it returns true.

https://llvm.org/doxygen/classllvm_1_1Linker.html#a5a864cc4d71e0234e9f8d786f8716804

Reply:
I changed it to be using the static `linkModules` function, they do the same thing but `linkModules` feels more intuitive to me :) 
:point_right: 4a77e9548af93922b53d6e8b8a5725d99dd8e418

Reply:
I changed the `linkInModule` function to use `Linker::Flags::OverrideFromSrc` it helps with resolving conflicts, for example when we don't want to clone the functions since its address has been taken.
Additionally, it removed the need to handle globals as they will be overridden based on source definition.

Reply:
In which commit?

Reply:
This one :point_right: https://github.com/ykjit/ykllvm/pull/202/commits/54bc3b9cfa2db6798c234db36937d9d256c620c4

Tests updated here :point_right: https://github.com/ykjit/ykllvm/pull/202/commits/6930afa62f29080b198ec4a32f699c390420cbf4